File: Samples/MapScenariosObjC/MapScenariosObjC/Scenarios/SearchPOIViewController.m
//
//  SearchPOIViewController.m
//  MapScenariosObjC
//
//  Created on 3/8/18.
//  Copyright © 2018 Phunware. All rights reserved.
//

#import <PWMapKit/PWMapKit.h>
#import <PWCore/PWCore.h>
#import <SDWebImage/UIImageView+WebCache.h>

#import "SearchPOIViewController.h"

static NSString * const poiCellReuseIdentifier = @"POICell";

@interface POITableViewCell: UITableViewCell

@property (nonatomic, strong) UIImageView *poiImageView;
@property (nonatomic, strong) UILabel *titleLabel;

- (void)configureSubviews;

@end

@implementation POITableViewCell

- (void)configureSubviews {
    if (self.poiImageView != nil && self.titleLabel != nil) {
        return;
    }
    
    self.poiImageView = [UIImageView new];
    self.titleLabel = [UILabel new];
    
    self.poiImageView.translatesAutoresizingMaskIntoConstraints = NO;
    [self addSubview:self.poiImageView];
    [self.poiImageView.leadingAnchor constraintEqualToAnchor:self.leadingAnchor constant:10.0].active = YES;
    [self.poiImageView.centerYAnchor constraintEqualToAnchor:self.centerYAnchor].active = YES;
    [self.poiImageView.widthAnchor constraintEqualToConstant:32.0].active = YES;
    [self.poiImageView.heightAnchor constraintEqualToConstant:32.0].active = YES;
    
    self.titleLabel.translatesAutoresizingMaskIntoConstraints = NO;
    self.titleLabel.numberOfLines = 0;
    [self addSubview:self.titleLabel];
    [self.titleLabel.leadingAnchor constraintEqualToAnchor:self.poiImageView.trailingAnchor constant:10.0].active = YES;
    [self.titleLabel.trailingAnchor constraintEqualToAnchor:self.trailingAnchor constant:5.0].active = YES;
    [self.titleLabel.topAnchor constraintEqualToAnchor:self.topAnchor constant:15.0].active = YES;
    [self.titleLabel.bottomAnchor constraintEqualToAnchor:self.bottomAnchor constant:-15.0].active = YES;
}

@end

@interface SearchPOIViewController () <UITableViewDataSource, UITableViewDelegate, UISearchResultsUpdating, UISearchBarDelegate>

@property (nonatomic, strong) NSString *applicationId;
@property (nonatomic, strong) NSString *accessKey;
@property (nonatomic, strong) NSString *signatureKey;

@property (nonatomic, strong) PWMapView *mapView;

@property (nonatomic, strong) UITableView *tableView;
@property (nonatomic, strong) NSArray *sortedPointsOfInterest;
@property (nonatomic, strong) NSArray *filteredPointsOfInterest;
@property (nonatomic, strong) UISearchController *searchController;

@end

@implementation SearchPOIViewController

- (id)initWithCoder:(NSCoder *)aDecoder {
    if (self = [super initWithCoder:aDecoder]) {
        _buildingIdentifier = 0; // Enter your building identifier here, found on the building's Edit page on Maas portal
        // Enter your application identifier, access key, and signature key, found on Maas portal under Account > Apps
        _applicationId = @"";
        _accessKey = @"";
        _signatureKey = @"";
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.navigationItem.title = @"Search for Point of Interest";
    
    if (self.applicationId.length > 0 && self.accessKey.length > 0 && self.signatureKey.length > 0) {
        [PWCore setApplicationID:self.applicationId accessKey:self.accessKey signatureKey:self.signatureKey];
    }
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(keyboardWillShow:) name:UIKeyboardWillShowNotification object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(keyboardWillHide:) name:UIKeyboardWillHideNotification object:nil];
    
    self.mapView = [PWMapView new];
    [self.view addSubview:self.mapView];
    [self configureMapViewConstraints];
    
    [PWBuilding buildingWithIdentifier:self.buildingIdentifier completion:^(PWBuilding *building, NSError *error) {
        __weak typeof(self) weakSelf = self;
        [weakSelf.mapView setBuilding:building animated:YES onCompletion:^(NSError *error) {
            dispatch_async(dispatch_get_main_queue(), ^{
                [weakSelf configureTableView];
            });
        }];
    }];
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)configureMapViewConstraints {
    self.mapView.translatesAutoresizingMaskIntoConstraints = NO;
    [[self.mapView.topAnchor constraintEqualToAnchor:self.view.topAnchor] setActive:YES];
    [[self.mapView.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor] setActive:YES];
    [[self.mapView.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor] setActive:YES];
    [[self.mapView.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor] setActive:YES];
}

- (void)configureTableView {
    NSSortDescriptor *sortDescriptor = [NSSortDescriptor sortDescriptorWithKey:@"title" ascending:YES];
    self.sortedPointsOfInterest = [self.mapView.building.pois sortedArrayUsingDescriptors:@[sortDescriptor]];
    self.filteredPointsOfInterest = [self.sortedPointsOfInterest copy];
    
    [self configureSearchController];
    self.tableView = [UITableView new];
    self.tableView.hidden = YES;
    [self.tableView registerClass:[POITableViewCell class] forCellReuseIdentifier:poiCellReuseIdentifier];
    self.tableView.rowHeight = UITableViewAutomaticDimension;
    self.tableView.dataSource = self;
    self.tableView.delegate = self;
    [self.view addSubview:self.tableView];
    [self configureTableViewConstraints];
}

- (void)configureTableViewConstraints {
    self.tableView.translatesAutoresizingMaskIntoConstraints = NO;
    [[self.tableView.topAnchor constraintEqualToAnchor:self.navigationController.navigationBar.bottomAnchor] setActive:YES];
    [[self.tableView.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor] setActive:YES];
    [[self.tableView.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor] setActive:YES];
    [[self.tableView.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor] setActive:YES];
}

- (void)configureSearchController {
    self.searchController = [[UISearchController alloc] initWithSearchResultsController:nil];
    self.searchController.searchResultsUpdater = self;
    self.searchController.obscuresBackgroundDuringPresentation = NO;
    self.searchController.searchBar.placeholder = @"Search Points of Interest";
    self.searchController.searchBar.delegate = self;
    self.definesPresentationContext = YES;
    self.navigationItem.searchController = self.searchController;
}

#pragma mark - UITableViewDataSource

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.filteredPointsOfInterest.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    POITableViewCell *poiCell = [tableView dequeueReusableCellWithIdentifier:poiCellReuseIdentifier forIndexPath:indexPath];
    [poiCell configureSubviews];
    
    PWPointOfInterest *pointOfInterest = self.filteredPointsOfInterest[indexPath.row];
    [poiCell.poiImageView sd_setImageWithURL:pointOfInterest.imageURL];
    poiCell.titleLabel.text = pointOfInterest.title;
    
    return poiCell;
}

#pragma mark - UITableViewDelegate

-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    [self.searchController.searchBar endEditing:YES];
    self.tableView.hidden = YES;
    PWPointOfInterest *pointOfInterest = self.filteredPointsOfInterest[indexPath.row];
    if (self.mapView.currentFloor.floorID != pointOfInterest.floorID) {
        PWFloor *newFloor = [self.mapView.building floorById:pointOfInterest.floorID];
        self.mapView.currentFloor = newFloor;
    }
    [self.mapView selectAnnotation:pointOfInterest animated:YES];
}

#pragma mark - UISearchResultsUpdating

- (void)updateSearchResultsForSearchController:(UISearchController *)searchController {
    if (searchController.searchBar.text == nil || searchController.searchBar.text.length == 0) {
        self.filteredPointsOfInterest = [self.sortedPointsOfInterest copy];
        [self.tableView reloadData];
        return;
    }
    
    NSPredicate *predicate = [NSPredicate predicateWithFormat:@"title contains[c] %@", searchController.searchBar.text];
    self.filteredPointsOfInterest = [self.sortedPointsOfInterest filteredArrayUsingPredicate:predicate];
    [self.tableView reloadData];
}

#pragma mark - UISearchBarDelegate

- (void)searchBarTextDidBeginEditing:(UISearchBar *)searchBar {
    self.tableView.hidden = NO;
}

- (void)searchBarTextDidEndEditing:(UISearchBar *)searchBar {
    self.tableView.hidden = YES;
}

#pragma mark - Adjust for keyboard

- (void)keyboardWillShow:(NSNotification *)notification {
    NSDictionary *userInfo = [notification userInfo];
    CGSize keyboardSize = [[userInfo objectForKey:UIKeyboardFrameBeginUserInfoKey] CGRectValue].size;
    
    self.tableView.contentInset = UIEdgeInsetsMake(0, 0, keyboardSize.height, 0);
    self.tableView.scrollIndicatorInsets = UIEdgeInsetsMake(0, 0, keyboardSize.height, 0);
}

- (void)keyboardWillHide:(NSNotification *)notification {
    self.tableView.contentInset = UIEdgeInsetsZero;
    self.tableView.scrollIndicatorInsets = UIEdgeInsetsZero;
}

@end

File: Samples/MapScenarios/MapScenarios/Scenarios/SearchPOIViewController.swift
//
//  SearchPOIViewController.swift
//  MapScenarios
//
//  Created on 3/7/18.
//  Copyright © 2018 Phunware. All rights reserved.
//

import Foundation
import UIKit
import PWMapKit
import PWCore

// MARK: - SearchPOIViewController
class SearchPOIViewController: UIViewController, ScenarioProtocol {
    
    // Enter your application identifier, access key, and signature key, found on Maas portal under Account > Apps
    var applicationId = ""
    var accessKey = ""
    var signatureKey = ""
    
    // Enter your campus identifier here, found on the campus's Edit page on Maas portal
    var campusIdentifier = 0

    // Enter your building identifier here, found on the building's Edit page on Maas portal
    var buildingIdentifier = 0
    
    private let mapView = PWMapView()
    
    // Search view
    private let tableView = UITableView()
    private let poiCellReuseIdentifier = "POICell"
    private var pointsOfInterest = [PWPointOfInterest]()
    private var filteredPointsOfInterest = [PWPointOfInterest]()
    private let searchController = UISearchController(searchResultsController: nil)
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        navigationItem.title = "Search for Points of Interest"
        
        if !validateScenarioSettings() {
            return
        }
        
        PWCore.setApplicationID(applicationId, accessKey: accessKey)
        
        NotificationCenter.default.addObserver(self, selector: #selector(keyboardWillShow(notification:)), name: UIResponder.keyboardWillShowNotification, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(keyboardWillHide(notification:)), name: UIResponder.keyboardWillHideNotification, object: nil)
        
        mapView.delegate = self
        view.addSubview(mapView)
        configureMapViewConstraints()
        searchController.searchBar.barStyle = .black
        navigationItem.searchController = searchController
        addFloorPickerView()
                
        // If we want to route between buildings on a campus, then we use PWCampus.campus to configure MapView
        // Otherwise, we will use PWBuilding.building to route between floors in a single building.
        if campusIdentifier != 0 {
            PWCampus.campus(identifier: campusIdentifier) { [weak self] result in
                switch result {
                case .success(let campus):
                    self?.mapView.setCampus(campus, animated: true, onCompletion: { (error) in
                        DispatchQueue.main.async {
                            self?.configureTableView()
                        }
                    })
                    
                case .failure(let error):
                    self?.warning(error.localizedDescription)
                }
            }
        }
        else {
            PWBuilding.building(identifier: buildingIdentifier) { [weak self] result in
                switch result {
                case .success(let building):
                    self?.mapView.setBuilding(building, animated: true, onCompletion: { (error) in
                        DispatchQueue.main.async {
                            self?.configureTableView()
                        }
                    })
                    
                case .failure(let error):
                    self?.warning(error.localizedDescription)
                }
            }
        }
    }
    
    override func viewDidAppear(_ animated: Bool) {
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self)
    }
    
    func configureMapViewConstraints() {
        mapView.translatesAutoresizingMaskIntoConstraints = false
        mapView.topAnchor.constraint(equalTo: view.topAnchor).isActive = true
        mapView.bottomAnchor.constraint(equalTo: view.bottomAnchor).isActive = true
        mapView.leadingAnchor.constraint(equalTo: view.leadingAnchor).isActive = true
        mapView.trailingAnchor.constraint(equalTo: view.trailingAnchor).isActive = true
    }
    
    func configureTableView() {
        guard let pois = mapView.pois() else {
            return
        }
        pointsOfInterest.append(contentsOf: pois)
        filteredPointsOfInterest = pointsOfInterest
        configureSearchController()
        tableView.isHidden = true
        tableView.register(POITableViewCell.self, forCellReuseIdentifier: poiCellReuseIdentifier)
        tableView.rowHeight = UITableView.automaticDimension
        tableView.dataSource = self
        tableView.delegate = self
        view.addSubview(tableView)
        configureTableViewConstraints()
    }
    
    func configureTableViewConstraints() {
        tableView.translatesAutoresizingMaskIntoConstraints = false
        tableView.topAnchor.constraint(equalTo: navigationController!.navigationBar.bottomAnchor).isActive = true
        tableView.bottomAnchor.constraint(equalTo: view.bottomAnchor).isActive = true
        tableView.leadingAnchor.constraint(equalTo: view.leadingAnchor).isActive = true
        tableView.trailingAnchor.constraint(equalTo: view.trailingAnchor).isActive = true
    }
    
    func configureSearchController() {
        searchController.searchResultsUpdater = self
        searchController.obscuresBackgroundDuringPresentation = false
        searchController.searchBar.placeholder = "Search Points of Interest"
        searchController.searchBar.delegate = self
    }
    
    func addFloorPickerView() {
        let bundleName = String(describing: FloorPickerView.self)
        let floorPickerView = Bundle.main.loadNibNamed(bundleName, owner: nil, options: nil)!.first as! FloorPickerView
        floorPickerView.configureInView(view, withMapView: mapView)
    }
}

// MARK: - UITableViewDataSource
extension SearchPOIViewController: UITableViewDataSource {
    
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return filteredPointsOfInterest.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let poiCell = tableView.dequeueReusableCell(withIdentifier: poiCellReuseIdentifier, for: indexPath) as! POITableViewCell
        poiCell.configureSubviews()
        
        let pointOfInterest = filteredPointsOfInterest[indexPath.row]
        poiCell.poiImageView.image = pointOfInterest.image
        if campusIdentifier != 0 {
            if let floorName = pointOfInterest.floor?.name,
               let buildingName = pointOfInterest.floor?.building.name,
               let poiName = pointOfInterest.title {
                poiCell.titleLabel.text = "\(buildingName):: \(floorName):: \(poiName)"
            }
        }
        else {
            if let floorName = pointOfInterest.floor?.name,
               let poiName = pointOfInterest.title {
                poiCell.titleLabel.text = "\(floorName):: \(poiName)"
            }
        }
        return poiCell
    }
}

// MARK: - PWMapViewDelegate
extension SearchPOIViewController: PWMapViewDelegate {
    
    func mapView(_ mapView: PWMapView!, didChange floor: PWFloor!) {
        mapView.zoomToFitFloor(floor)
    }
}

// MARK: - UITableViewDelegate
extension SearchPOIViewController: UITableViewDelegate {
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        searchController.searchBar.endEditing(true)
        self.tableView.isHidden = true
        let pointOfInterest = filteredPointsOfInterest[indexPath.row]
        if mapView.currentFloor.floorID != pointOfInterest.floorID {
            let newFloor = mapView.floorById(pointOfInterest.floorID)
            mapView.currentFloor = newFloor
        }
        mapView.selectAnnotation(pointOfInterest, animated: true)
    }
}

// MARK: - UISearchResultsUpdating
extension SearchPOIViewController: UISearchResultsUpdating {
    
    func updateSearchResults(for searchController: UISearchController) {
        guard let searchText = searchController.searchBar.text, searchText.count > 0 else {
            filteredPointsOfInterest = pointsOfInterest
            tableView.reloadData()
            return
        }
        filteredPointsOfInterest = pointsOfInterest.filter({( pointOfInterest : PWPointOfInterest) -> Bool in
            guard let title = pointOfInterest.title else {
                return false
            }
            return title.lowercased().contains(searchText.lowercased())
        })
        
        tableView.reloadData()
    }
}

// MARK: - UISearchBarDelegate
extension SearchPOIViewController: UISearchBarDelegate {
    
    func searchBarTextDidBeginEditing(_ searchBar: UISearchBar) {
        tableView.isHidden = false
    }
    
    func searchBarTextDidEndEditing(_ searchBar: UISearchBar) {
        tableView.isHidden = true
    }
}

// MARK: - Adjust for keyboard
extension SearchPOIViewController {
    
    @objc func keyboardWillShow(notification: Notification) {
        if let keyboardSize = (notification.userInfo?[UIResponder.keyboardFrameBeginUserInfoKey] as? NSValue)?.cgRectValue {
            tableView.contentInset = UIEdgeInsets(top: 0, left: 0, bottom: keyboardSize.height, right: 0)
            tableView.scrollIndicatorInsets = UIEdgeInsets(top: 0, left: 0, bottom: keyboardSize.height, right: 0)
        }
    }
    
    @objc func keyboardWillHide(notification: Notification) {
        tableView.contentInset = .zero
        tableView.scrollIndicatorInsets = .zero
    }
}

// MARK: - POITableViewCell
class POITableViewCell: UITableViewCell {
    
    var poiImageView: UIImageView!
    var titleLabel: UILabel!
    
    func configureSubviews() {
        if poiImageView != nil && titleLabel != nil {
            return
        }
        
        poiImageView = UIImageView()
        titleLabel = UILabel()
        
        poiImageView.translatesAutoresizingMaskIntoConstraints = false
        addSubview(poiImageView)
        poiImageView.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 10.0).isActive = true
        poiImageView.centerYAnchor.constraint(equalTo: centerYAnchor).isActive = true
        poiImageView.widthAnchor.constraint(equalToConstant: 32.0).isActive = true
        poiImageView.heightAnchor.constraint(equalToConstant: 32.0).isActive = true
        
        titleLabel.translatesAutoresizingMaskIntoConstraints = false
        titleLabel.numberOfLines = 0
        addSubview(titleLabel)
        titleLabel.leadingAnchor.constraint(equalTo: poiImageView.trailingAnchor, constant: 10.0).isActive = true
        titleLabel.trailingAnchor.constraint(equalTo: trailingAnchor, constant: 5.0).isActive = true
        titleLabel.topAnchor.constraint(equalTo: topAnchor, constant: 15.0).isActive = true
        titleLabel.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -15.0).isActive = true
    }
}
